### PR TITLE
[CI] Combine action templates with matrix

### LINKED ISSUE #178: Combine Duplicate CI jobs for backend & algorithm
## Feature Description

Implement [Matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) for the Backend & Algorithm CI jobs to reduce them down to two files

## Problem You're Trying to Solve

Each service has its own GitHub Workflow file which means a lot of duplication (7:2 to be exact) which is becoming annoying to manage and add new CI jobs to.

### PR DESCRIPTION
**Pull request by MNThomson**
# Description

Convert all the backend & algorithm jobs into two files using a matrix strategy

Closes #178

## How to Test

Observe all the green checkmarks in CIC

## Checklist
- [ ] The code includes tests if relevant
- [x] I have *actually* self-reviewed my changes and done QA

### CODE REVIEW

File: .github/workflows/backend.yml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+      - name: Docker Compose Up
+        run: docker compose up -d ${{ matrix.service == 'lib' && 'user' || matrix.service }} datalayer database
+
+      - name: Install Dependencies
+        if: matrix.service == 'lib'
+        run: docker compose exec user npm install --prefix ../lib
+
+      - name: Test
+        run: docker compose exec ${{ matrix.service == 'lib' && 'user' || matrix.service }} npm run ${{ matrix.service == 'lib' && '--prefix ../lib' || '' }} test:ci
```
Comment (kylestang) [inline .github/workflows/backend.yml:79]: :fearful:

File: apps/backend/lib/package.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "martletplace-listing",
+  "name": "martletplace-lib",
```
Comment (kylestang) [inline apps/backend/lib/package.json:2]: Who approved this???

**Review by kylestang — APPROVED** [review]
Looks cursed to me, but ok

File: apps/backend/lib/package.json (hunk 2 above)
Comment (MNThomson) [reply apps/backend/lib/package.json:2]: Who wrote this???

**Pull request merged**